dataservice/api/biospecimen/manager.py

```diff
@@ -0,0 +1,312 @@
+        biospecimen.age_at_event_days
+    ]
+
+    return "-".join([str(c) if c else "Not Reported" for c in components])
+
+
+def _create_concentration(biospecimen):
+    """
+    Create the sample concentration given the biospecimen concentration
+
+    Only use the concentration value when the analyte_type is DNA or RNA
+    """
+    if biospecimen.analyte_type in ["DNA", "RNA"]:
+        return biospecimen.concentration_mg_per_ml
+    else:
+        return None
+
+
+def _get_sample_identifier(biospecimen):
+    """
+    Helper to extract specific Biospecimen attributes to uniquely
+    identify a Sample
+    """
+    return {
+        "sample_event_key": _create_sample_event_key(biospecimen),
+        "composition": biospecimen.composition,
+        "tissue_type": biospecimen.source_text_tissue_type,
+        "analyte_type": biospecimen.analyte_type,
+        "anatomical_location": biospecimen.source_text_anatomical_site,
+        "method_of_sample_procurement":
+        biospecimen.method_of_sample_procurement,
+        "preservation_method": biospecimen.preservation_method,
+        "concentration_mg_per_ml": _create_concentration(biospecimen)
+    }
+
+
+def _get_container_identifier(biospecimen):
+    """
+    Helper to extract specific Biospecimen attributes to uniquely identify
+    a Container
+    """
+    return {
+        "biospecimen_id": biospecimen.kf_id,
+    }
+
+
+def _get_visibility_params(biospecimen):
+    """
+    Helper method to get dict of visibility parameters from the Biospecimen
+    """
+    return {
+        "visible": biospecimen.visible,
+        "visibility_reason": biospecimen.visibility_reason,
+        "visibility_comment": biospecimen.visibility_comment
+    }
+
+
+def _create_sample(biospecimen):
+    """
+    Create Sample from specific Biospecimen attributes. Validate Sample
+    """
+    # Extract the parameters that uniquely identify a sample
+    params = _get_sample_identifier(biospecimen)
+    # Add remaining sample attributes
+    params.update(
+        {
+            "participant_id": biospecimen.participant_id,
+            "external_id": biospecimen.external_sample_id,
+            "volume_ul": biospecimen.volume_ul,
+        }
+    )
+    # Set visibility params based on Biospecimen which represents both
+    # sample and containers
+    params.update(
+        _get_visibility_params(biospecimen)
+    )
+    # Validate sample parameters and create sample
+    try:
+        sample = SampleSchema(strict=True).load(params).data
+    # Params not valid
+    except ValidationError as e:
+        abort(400, 'could not create sample: {}'.format(e.messages))
+
+    return sample
+
+
+def _update_sample(current_sample, biospecimen):
+    """
+    Update Sample using specific Biospecimen attributes. Validate Sample
+    """
+    # Extract the parameters that uniquely identify a sample
+    params = _get_sample_identifier(biospecimen)
+    # Add remaining sample attributes
+    params.update(
+        {
```

Comment:
everything in this PR makes sense to me except the idea that we might update a participant ID.  I _think_ participant ID should be part of the defining characteristics of a sample so I'm struggling to understand how we could both identify an existing sample (which implies the participant ID on the sample matches that on the specimen being registered) but then update the sample participant ID field (which implies the participant ID does not match the specimen being registered).

Reply:
Oh hold on... is this related to participant.kf_id really being the primary ID for particpant and participant_id being a sort of secondary/external ID?  So we are updating the _external_ ID if it changes but relying on the kf_id/PK for confirming the sample already exists?

Reply:
@calkinsh  Yep, the primary key for participant is `participant.kf_id` and the sample has a foreign key to it `sample.participant_id` so I think that does make it a defining characteristic of the sample.

Reply:
Technically we may not need the `Sample.participant_id` or `Sample.external_id` bc they are captured in the `Sample.sample_event_key` but I included them in the Sample table in case we want to populate the sample event key with something and bc I felt it would be ok to have some redundancy to gain some clarity on which participant the sample came from and what the original biospecimen's external sample ID was